api/src/main/java/org/openmrs/module/reportingcompatibility/reporting/export/DataExportFunctions.java

```diff
@@ -1,1352 +1,1356 @@
+	
+	// Map<name of drug set, Map<patientId, List<DrugOrder>>>
+	protected Map<String, Map<Integer, List<DrugOrder>>> drugOrderMap = new HashMap<String, Map<Integer, List<DrugOrder>>>();
+	
+	// Map<name of drug set, Map<patientId, List<DrugOrder>>>
+	protected Map<String, Map<Integer, List<DrugOrder>>> currentDrugOrderMap = new HashMap<String, Map<Integer, List<DrugOrder>>>();
+	
+	// Map<tablename+columnname, Map<patientId, columnvalue>>
+	protected Map<String, Map<Integer, Object>> patientAttributeMap = new HashMap<String, Map<Integer, Object>>();
+	
+	// Map<tablename+columnname, Map<personId, columnvalue>>
+	protected Map<String, Map<Integer, Object>> personAttributeMap = new HashMap<String, Map<Integer, Object>>();
+	
+	// Map<key, Collection<personId>>, where key is like "Cohort.1" or "Filter.3"
+	protected Map<String, Collection<Integer>> cohortMap = new HashMap<String, Collection<Integer>>();
+	
+	protected ReportingCompatibilityService rcs;
+	
+	protected PatientService patientService;
+	
+	protected PersonService personService;
+	
+	protected ConceptService conceptService;
+	
+	protected EncounterService encounterService;
+	
+	protected Locale locale = null;
+	
+	// Constructors
+	
+	public DataExportFunctions(Patient p) {
+		this(p.getPatientId());
+	}
+	
+	public DataExportFunctions(Integer patientId) {
+		this();
+		setPatientId(patientId);
+	}
+	
+	public DataExportFunctions() {
+		this.rcs = Context.getService(ReportingCompatibilityService.class);
+		this.patientService = Context.getPatientService();
+		this.conceptService = Context.getConceptService();
+		this.encounterService = Context.getEncounterService();
+		this.personService = Context.getPersonService();
+		
+		locale = Context.getLocale();
+		dateFormatLong = DateFormat.getDateTimeInstance(DateFormat.LONG, DateFormat.LONG, locale);
+		dateFormatShort = Context.getDateFormat();
+		dateFormatYmd = new SimpleDateFormat("yyyy-MM-dd", locale);
+	}
+	
+	public void clearAllMaps() {
+		clearMap(patientEncounterMap);
+		clearMap(patientIdentifierMap);
+		clearMap(patientFirstEncounterMap);
+		conceptNameMap.clear();
+		clearMap(conceptAttrObsMap);
+		clearMap(relationshipMap);
+		clearMap(programMap);
+		clearMap(drugOrderMap);
+		clearMap(currentDrugOrderMap);
+		clearMap(patientAttributeMap);
+		clearMap(personAttributeMap);
+		cohortMap.clear();
+	}
+		
+	public void clear() {
+		clearAllMaps();
+		rcs = null;
+		patientService = null;
+		conceptService = null;
+		encounterService = null;
+		personService = null;
+	}
+	
+	@SuppressWarnings("unchecked")
+	private void clearMap(Map<?, ? extends Map> mapOfMaps) {
+		for (Map map : mapOfMaps.values())
+			map.clear();
+		mapOfMaps.clear();
+	}
+	
+	/**
+	 * Called when garbage collecting this class
+	 * 
+	 * @see java.lang.Object#finalize()
+	 */
+	@Override
+	protected void finalize() throws Throwable {
+		if (log.isDebugEnabled())
+			log.debug("GC is collecting the data export functions..." + this);
+		super.finalize();
```

Comment:
Our static analysis engine detects that your pull request includes a change of **explicit invocation of finalizer**. According to [Java doc](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#finalize--), finalizer methods are designed **to be executed once, and only by the Virtual Machine (VM)**. If a connected set of objects beings finalizable, then the VM will invoke the finalize method on all the finalizable object, possibly at the same time in different threads. Thus, it is a particularly bad idea, in the finalize method for a class X, invoke finalize on objects referenced by X, because they may already be getting finalized in a separate thread.